prov/gni/test/vector.c

```diff
@@ -0,0 +1,348 @@
+{
+	int ret;
+	void *tmp = malloc(sizeof(gnix_vec_entry_t));
+	cr_assert(tmp, "do_insert_first");
+
+	ret = vec.ops->insert_first(&vec, tmp);
+	cr_assert(!ret, "_gnix_vec_insert_first");
+
+	ret = vec.ops->insert_first(&vec, tmp);
+	cr_assert_eq(ret, -FI_ECANCELED);
+}
+
+void do_insert_last()
+{
+	int ret;
+	void *tmp = malloc(sizeof(gnix_vec_entry_t));
+	cr_assert(tmp, "do_insert_last");
+
+	ret = vec.ops->insert_last(&vec, tmp);
+	cr_assert(!ret, "_gnix_vec_insert_last");
+
+	ret = vec.ops->insert_last(&vec, tmp);
+	cr_assert_eq(ret, -FI_ECANCELED);
+}
+
+void do_fill_insert_at()
+{
+	int i, ret;
+	void *tmp;
+
+	for (i = 0; i < vec.attr.cur_size; i++) {
+		tmp = malloc(sizeof(gnix_vec_entry_t));
+		cr_assert(tmp, "do_insert_at");
+
+		ret = vec.ops->insert_at(&vec, tmp, i);
+		cr_assert(!ret, "_gnix_vec_insert_at");
+	}
+
+	/* Test grow. */
+	tmp = malloc(sizeof(gnix_vec_entry_t));
+	cr_assert(tmp, "do_insert_at");
+
+	ret = vec.ops->insert_at(&vec, tmp, VEC_MAX-1);
+	cr_assert(!ret, "_gnix_vec_insert_at");
+	cr_assert_eq(vec.attr.cur_size, VEC_MAX);
+
+	ret = vec.ops->insert_at(&vec, tmp, VEC_MAX-1);
+	cr_assert_eq(ret, -FI_ECANCELED);
+
+	for (; i < vec.attr.cur_size - 1; i++) {
+		tmp = malloc(sizeof(gnix_vec_entry_t));
+		cr_assert(tmp, "do_insert_at");
+
+		ret = vec.ops->insert_at(&vec, tmp, i);
+		cr_assert(!ret, "_gnix_vec_insert_at");
+	}
+
+	ret = vec.ops->insert_at(&vec, tmp, VEC_MAX);
+	cr_assert_eq(ret, -FI_EINVAL);
+}
+
+void do_remove_first()
+{
+	int ret;
+
+	ret = vec.ops->remove_first(&vec);
+	cr_assert(!ret, "_gnix_vec_remove_first");
+
+	ret = vec.ops->remove_first(&vec);
+	cr_assert_eq(ret, -FI_ECANCELED);
+}
+
+void do_remove_last()
+{
+	int ret;
+
+	ret = vec.ops->remove_last(&vec);
+	cr_assert(!ret, "_gnix_vec_remove_last");
+
+	ret = vec.ops->remove_last(&vec);
+	cr_assert_eq(ret, -FI_ECANCELED);
+}
+
+void do_unfill_remove_at()
+{
+	int i, ret;
+
+	for (i = 0; i < vec.attr.cur_size; i++) {
+		ret = vec.ops->remove_at(&vec, i);
+		cr_assert(!ret, "_gnix_vec_remove_at");
+	}
+
+	ret = vec.ops->remove_at(&vec, vec.attr.cur_size / 2);
+	cr_assert_eq(ret, -FI_ECANCELED);
+}
+
+void do_first()
+{
+	int ret;
+	void *tmp;
+
+	ret = vec.ops->first(&vec, &tmp);
+	cr_assert(tmp, "_gnix_vec_first");
+	cr_assert(!ret, "_gnix_vec_first");
+
+	tmp = NULL;
+
+	do_remove_first();
+
+	ret = vec.ops->first(&vec, &tmp);
+	cr_assert_eq(tmp, NULL);
+	cr_assert_eq(ret, -FI_ECANCELED);
+}
+
+void do_last()
+{
+	int ret;
+	void *tmp;
+
+	ret = vec.ops->last(&vec, &tmp);
+	cr_assert(tmp, "_gnix_vec_last");
+	cr_assert(!ret, "_gnix_vec_last");
+
+	tmp = NULL;
+
+	do_remove_last();
+
+	ret = vec.ops->last(&vec, &tmp);
+	cr_assert_eq(tmp, NULL);
+	cr_assert_eq(ret, -FI_ECANCELED);
+}
+
+void do_at()
+{
+	int i ,ret;
+	void *tmp;
+
+	for (i = 0; i < vec.attr.cur_size; i++) {
+		ret = vec.ops->at(&vec, &tmp, i);
+		cr_assert(!ret, "_gnix_vec_at");
+
+		cr_assert(!!tmp, "_gnix_vec_at");
+		tmp = NULL;
+	}
+}
+
+void do_tests()
```

Comment:
Each of these should probably be their own criterion test for the sake of granularity.